### PR TITLE
Update LogMonitoring validation logic

### PR DESCRIPTION
**Pull request by 0sewa0**
## Description
[DAQ-1430](https://dt-rnd.atlassian.net/browse/DAQ-1430) (has pictures for better understanding of supported scenarios)

Updates the validation logic (and some flow-control logic) related to how we handle the `LogMonitoring` section.

Main change: [Add warning for LogMonitoring TemplateSpec being ignored](https://github.com/Dynatrace/dynatrace-operator/commit/252208997b624107e683459d42cfe910001e4504)
- `LogMonitoring` and `OneAgent` can be enabled at the same time (so no longer denied)
- BUT, the `LogMonitoring`'s Template section will be ignored, and we will print a warning and only the OneAgent DaemonSet is deployed.

Check commits for details.
- Some changes were needed, to make the warning possible while not breaking the Operator's logic
  -  [Add IsStandalone check for LogMonitoring](https://github.com/Dynatrace/dynatrace-operator/commit/74d6737575e47e7d89c74ba3c194486fe80f279a)
     - So we don't deploy multiple DaemonSets
   - [Make LogMonitoring TemplateSpec nilable](https://github.com/Dynatrace/dynatrace-operator/commit/6b4b10cba88f1979bbd7a950fdcd7fbd87b62ab1)
      - So we know that the TemplateSpec was set or not 

## How can this be tested?

**The following is allowed without warnings:**
```
apiVersion: dynatrace.com/v1beta3
kind: DynaKube
metadata:
  name: dk
  namespace: dynatrace
spec:
  apiUrl: https://<tenant>.dev.dynatracelabs.com/api
  logMonitoring: {}
  oneAgent:
    cloudNativeFullStack: {}
```
- Only the OneAgent DaemonSet should be deployed
- The LogMonitoring's config-secret shouldn't be created
- No LogMonitoring related conditions (as of right now) on the DynaKube

**The following is allowed with warnings:**
```
apiVersion: dynatrace.com/v1beta3
kind: DynaKube
metadata:
  name: dk
  namespace: dynatrace
spec:
  apiUrl: https://<tenant>.dev.dynatracelabs.com/api
  logMonitoring: {}
  oneAgent:
    cloudNativeFullStack: {}
  templates:
    logMonitoring:
      nodeSelector:
        dk-log: deploy
```
- The `templates.logMonitoring.nodeSelector` is totally ignored (the whole `templates.logMonitoring` is ignored)
   - During the DaemonSet generation and `conflictingNodeSelector` validation as well 

[DAQ-1430]: https://dt-rnd.atlassian.net/browse/DAQ-1430?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: pkg/api/validation/dynakube/logmonitoring.go [hunk 1]
```diff
@@ -2,52 +2,21 @@ package validation
...
 
 	"github.com/Dynatrace/dynatrace-operator/pkg/api/v1beta3/dynakube"
-	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 const (
-	errorConflictingLogMonitoring = "The DynaKube's specification tries to enable LogMonitoring in a namespace where another DynaKube already deploys the OneAgent, which is not supported. The conflicting DynaKubes: %s"
-
-	errorConflictingOneAgentSpec = "The DynaKube's specification tries to enable LogMonitoring and OneAgent at the same time, which is not supported. Please disable the LogMonitoring or OneAgent"
+	warningLogMonitoringIgnoredTemplate = "The DynaKube's specification tries to configure LogMonitoring Template section and OneAgent at the same time, in which case the LogMonitoring Template section is ignored"
```
Comment (andriisoldatenko) [inline pkg/api/validation/dynakube/logmonitoring.go:10]:
```
and OneAgent at the same time, in which case the LogMonitoring Template
```
i'm a bit confused with this sentence, do you want to say: "and OneAgent at the same time, in *this* case the LogMonitoring Template" ? or am I read it incorrectly?

File: pkg/api/validation/dynakube/logmonitoring.go [hunk 2]
```diff
@@ -2,52 +2,21 @@ package validation
...
 )
 
-func conflictingLogMonitoringNodeSelector(ctx context.Context, dv *Validator, dk *dynakube.DynaKube) string {
-	if !dk.LogMonitoring().IsEnabled() {
+func ignoredLogMonitoringTemplates(ctx context.Context, dv *Validator, dk *dynakube.DynaKube) string {
```
Comment (andriisoldatenko) [inline pkg/api/validation/dynakube/logmonitoring.go:13]:
```suggestion
func ignoredLogMonitoringTemplate(ctx context.Context, dv *Validator, dk *dynakube.DynaKube) string {
```

File: pkg/api/validation/dynakube/logmonitoring_test.go [hunk 3]
```diff
@@ -1,59 +1,27 @@
...
 
-const (
-	testNodeSelector     = "nodeselector"
-	testCnfsDynakubeName = "CNFS"
-)
-
-func TestConflictingLogMonitoringConfiguration(t *testing.T) {
-	t.Run("oneagent and log module not possible on the same not at the same time in one dk", func(t *testing.T) {
-		dk := createLogMonitoringDynakube(testName, "")
+func TestIgnoredLogMonitoringTemplates(t *testing.T) {
```
Comment (andriisoldatenko) [inline pkg/api/validation/dynakube/logmonitoring_test.go:11]:
```suggestion
func TestIgnoredLogMonitoringTemplate(t *testing.T) {
```

File: pkg/api/validation/dynakube/oneagent.go [hunk 4]
```diff
@@ -18,7 +18,7 @@ const (
 
 	errorImageFieldSetWithoutCSIFlag = `The DynaKube specification attempts to enable ApplicationMonitoring mode and retrieve the respective image, but the CSI driver is not enabled.`
 
-	errorNodeSelectorConflict = `The DynaKube specification attempts to deploy a OneAgent, which conflicts with another DynaKube's OneAgent/LogMonitoring. Only one Agent per node is supported.
+	errorNodeSelectorConflict = `The DynaKube specification attempts to deploy a OneAgent/LogMonitoring, which conflicts with another DynaKube's OneAgent/LogMonitoring. Only one Agent per node is supported.
```
Comment (andriisoldatenko) [inline pkg/api/validation/dynakube/oneagent.go:21]:
```suggestion
	errorNodeSelectorConflict = `The DynaKube specification attempts to deploy a OneAgent/LogMonitoring, which conflicts with another DynaKube's OneAgent or LogMonitoring. Only one Agent per node is supported.
```

File: pkg/api/validation/dynakube/logmonitoring.go (hunk 1 above)
Comment (chrismuellner) [reply pkg/api/validation/dynakube/logmonitoring.go:10]: the wording sounds correct to me, you could change "OneAgent" to "OneAgent section" or "OneAgent section using host agent" or similar to make it clearer

Comment (0sewa0) [reply pkg/api/validation/dynakube/logmonitoring.go:10]: [example usage of "in which case"](https://www.merriam-webster.com/dictionary/in%20which%20case)

Comment (andriisoldatenko) [reply pkg/api/validation/dynakube/logmonitoring.go:10]: okey to be super clear: 

image

☝️  this 2 together - read very strangly. I tried to translate it and either you forgot to add period or I miss something.

Comment (0sewa0) [reply pkg/api/validation/dynakube/logmonitoring.go:10]: I already came up with an IMO better way, so don't worry

Comment (andriisoldatenko) [reply pkg/api/validation/dynakube/logmonitoring.go:10]: > The DynaKube's specification tries to configure LogMonitoring Template section and OneAgent together, in which case, the LogMonitoring Template section is ignored.

IMO much better to read

File: pkg/api/v1beta3/dynakube/logmonitoring/props.go [hunk 5]
```diff
@@ -1,17 +1,34 @@
...
+
+func (lm *LogMonitoring) GetDaemonSetName() string {
+	return lm.name + daemonSetSuffix
+}
+
+func (lm *LogMonitoring) IsStandalone() bool {
+	return lm.IsEnabled() && !lm.enabledDependencies.hostAgents
+}
+
+// Template is a nil-safe way to access the underlying TemplateSpec
```
Comment (andriisoldatenko) [inline pkg/api/v1beta3/dynakube/logmonitoring/props.go:27]:
```suggestion
// Template is a nil-safe way to access the underlying TemplateSpec.
```

File: pkg/api/validation/dynakube/logmonitoring.go (hunk 1 above)
Comment (chrismuellner) [reply pkg/api/validation/dynakube/logmonitoring.go:10]:
```suggestion
	warningLogMonitoringIgnoredTemplate = "The Dynakube's specification skips LogMonitoring Template section as OneAgent section is configured at the same time."
```

Comment (andriisoldatenko) [reply pkg/api/validation/dynakube/logmonitoring.go:10]: 👍

**Review by waodim — APPROVED** [review]

File: pkg/api/validation/dynakube/oneagent.go [hunk 6]
```diff
@@ -60,7 +60,7 @@ func conflictingOneAgentConfiguration(_ context.Context, _ *Validator, dk *dynak
 }
 
 func conflictingOneAgentNodeSelector(ctx context.Context, dv *Validator, dk *dynakube.DynaKube) string {
-	if !dk.NeedsOneAgent() || dk.FeatureEnableMultipleOsAgentsOnNode() {
+	if (!dk.LogMonitoring().IsStandalone() && !dk.NeedsOneAgent()) || dk.FeatureEnableMultipleOsAgentsOnNode() {
```
Comment (chrismuellner) [inline pkg/api/validation/dynakube/oneagent.go:63]: I'm not sure the feature flag is valid for log agent as well

File: pkg/api/validation/dynakube/oneagent.go [hunk 7]
```diff
@@ -60,7 +60,7 @@ func conflictingOneAgentConfiguration(_ context.Context, _ *Validator, dk *dynak
 }
 
 func conflictingOneAgentNodeSelector(ctx context.Context, dv *Validator, dk *dynakube.DynaKube) string {
```
Comment (chrismuellner) [inline pkg/api/validation/dynakube/oneagent.go:62]: shouldn't we validate the conflicting node selectors in a single location and not for both oneagent and log agent in a slightly different way? We can't have 2 of them on the same node, no matter if it's log agent or oneagent

Comment (0sewa0) [reply pkg/api/validation/dynakube/oneagent.go:62]: before the PR it was separated, but now it's in 1 function, sooo 😅

**Review by chrismuellner — APPROVED** [review]

**Pull request merged**